### PR TITLE
feat: check if stream matches input format via file signature/structure

### PR DESCRIPTION
**Pull request by ok-nick**
Adds extra validation before loading the JUMBF to confirm the input format type matches the given stream, otherwise outputting a more specific error.

Some format types such as MP3 and SVG may require a more in-depth structural check.

### CODE REVIEW

File: sdk/src/asset_handlers/c2pa_io.rs [hunk 1]
```diff
@@ -128,6 +128,11 @@ impl AssetIO for C2paIO {
         &SUPPORTED_TYPES
     }
 
+    fn supports_stream(&self, _stream: &mut dyn CAIRead) -> Result<bool> {
+        // TODO: complex
+        Ok(true)
```
Comment (mauricefisher64) [inline sdk/src/asset_handlers/c2pa_io.rs]: You can look for the C2PA JUMP header.

File: sdk/src/asset_handlers/svg_io.rs [hunk 2]
```diff
@@ -159,6 +159,11 @@ impl AssetIO for SvgIO {
     fn supported_types(&self) -> &[&str] {
         &SUPPORTED_TYPES
     }
+
+    fn supports_stream(&self, _stream: &mut dyn CAIRead) -> Result<bool> {
+        // TODO: complex
+        Ok(true)
```
Comment (mauricefisher64) [inline sdk/src/asset_handlers/svg_io.rs]: You could check for this "<?xml version"

Comment (ok-nick) [reply sdk/src/asset_handlers/svg_io.rs]: I'm looking at a few implementations online, and it seems libraries are reading the first ~4096 bytes and looking for the `<svg` string.

File: sdk/src/asset_handlers/svg_io.rs [hunk 3]
```diff
@@ -159,6 +159,15 @@ impl AssetIO for SvgIO {
     fn supported_types(&self) -> &[&str] {
         &SUPPORTED_TYPES
     }
+
+    fn supports_stream(&self, stream: &mut dyn CAIRead) -> Result<bool> {
+        stream.rewind()?;
+
+        let mut bytes = [0u8; 4096];
```
Comment (tmathern) [inline sdk/src/asset_handlers/svg_io.rs]: If we think more checks are needed, I would not add it here now. Either split it out, or do the full check. I'd rather we be less restrictive to possibly avoid breaking edge cases now - or due the complete due diligence.

@ok-nick What makes you say more checks may be needed for this case?

File: sdk/src/asset_handlers/mp3_io.rs [hunk 4]
```diff
@@ -344,6 +344,18 @@ impl AssetIO for Mp3IO {
...
     }
+
+    fn supports_stream(&self, stream: &mut dyn CAIRead) -> Result<bool> {
+        stream.rewind()?;
+
+        let mut header = [0u8; 3];
+        stream.read_exact(&mut header)?;
+
+        Ok(
+            (header[0] == 0x49 && header[1] == 0x44 && header[2] == 0x33)
```
Comment (tmathern) [inline sdk/src/asset_handlers/mp3_io.rs]: If we think more checks are needed, I would not add it here now. Either split it out, or do the fill check. I'd rather we be less restrictive to possibly avoid breaking edge cases now - or due the compelte due diligence.

@ok-nick What makes you say more checks may be needed for this case?

**Review by tmathern — CHANGES_REQUESTED** [review]
If we are unsure about mp3 and svg and need more work/verifications, I'd rather we split it out. To avoid taking the risk of breaking (currently working) edge cases.
The corollary is to add good test coverage using real files to raise confidence we covered all cases (is it even possible with mp3?).

File: sdk/src/asset_handlers/mp3_io.rs (hunk 4 above)
Comment (ok-nick) [reply sdk/src/asset_handlers/mp3_io.rs]: These functions are only used as an extra check to confirm the given stream can possibly be the specified format, if not error with a descriptive error. They are purposefully less-restrictive so that they don't block correct cases. For instance, we check [0x49, 0x44, 0x33] as the starting bytes for MP3, but that may only indicate an ID3 block which may or may not be MP3, we leave the actual parsing up to the parser.

Comment (tmathern) [reply sdk/src/asset_handlers/mp3_io.rs]: Ok!

File: sdk/src/asset_handlers/svg_io.rs (hunk 3 above)
Comment (ok-nick) [reply sdk/src/asset_handlers/svg_io.rs]: Hm, for SVG it would be more correct to parse the XML and check for a top-level `svg` tag, but that is quite expensive. Ideally we'd construct the parser once with a stream and we can do the check there as we parse it, then cache all the important info in one pass (this is related to Gavin's content credential proposal), but that's a much larger amount of work.

But to answer your question, this is the least-restrictive check, similar to MP3, as a pre-validation step.

**Comment by ok-nick** [comment]
I will be changing this PR to perform the support checks as it's being read. This will be much more reliable since we can use our normal parsing methods (like XML for SVG) and it can be done in one pass. We also do this already for some formats, like GIF.

File: sdk/src/asset_handlers/pdf_io.rs [hunk 5]
```diff
@@ -28,6 +27,21 @@ pub struct PdfIO {}
 impl CAIReader for PdfIO {
     fn read_cai(&self, asset_reader: &mut dyn CAIRead) -> crate::Result<Vec<u8>> {
         asset_reader.rewind()?;
+
+        let mut header = [0u8; 5];
+        asset_reader.read_exact(&mut header)?;
```
Comment (mauricefisher64) [inline sdk/src/asset_handlers/pdf_io.rs]: This is wrong.  The file is a text file so you could have had spaces or nulls before you get to the %PDF

File: sdk/src/error.rs [hunk 6]
```diff
@@ -370,6 +376,34 @@ pub enum Error {
     // The string should be one of the C2PA validation codes
     #[error("C2PA Validation Error: {0}")]
     C2PAValidation(String),
+
+    #[error("error parsing BMFF: {0}")]
```
Comment (mauricefisher64) [inline sdk/src/error.rs:380]: Do you plan on adding more error codes for each format specific error?  Seems odd to have made these with just a single error enum for each.

Comment (ok-nick) [reply sdk/src/error.rs:380]: Yes it is odd right now, I agree. Eventually I imagine that all error types returned by (for example) BMFF would be contained in the `BmffError` enum. Any error returned would then include the `error parsing BMFF: some more specific error` context. That would make it much easier to reason about adding more granular errors to the asset handlers without appending it onto the already giant `c2pa::Error` enum.

File: sdk/src/asset_handlers/pdf_io.rs (hunk 5 above)
Comment (ok-nick) [reply sdk/src/asset_handlers/pdf_io.rs]: I took this behavior from the wikipedia [here](https://en.wikipedia.org/wiki/List_of_file_signatures). Are you saying that for non-spec compliant PDFs or is that part of the spec?

Comment (mauricefisher64) [reply sdk/src/asset_handlers/pdf_io.rs]: I am saying that some PDFs could have space before the %PDF- and it is still valid

Comment (ok-nick) [reply sdk/src/asset_handlers/pdf_io.rs]: Discussed more offline. We decided that the PDF parser can determine if there is a problem and only do file signature checks if it's reliable to do an easy one.
